examples/llm/glem.py

```diff
@@ -0,0 +1,440 @@
+"""This example run GLEM model using PyG.
+Original Paper: https://arxiv.org/abs/2210.14709
+“Learning on Large-scale Text-attributed Graphs via Variational Inference“.
+Requirements on top of basic PyG:
+`pip install ogb transformers peft tqdm`.
+GLEM is a data augmentation co-training strategy for LM and GNN, our
+implementation extended original implementation from LM to LLM and opt for LoRA
+from peft.
+
+``note::
+    use addtional trick, please add your external prediction by assigning
+    `ext_pred_path` and combine it into pretraining phase and node features
+"""
+
+import argparse
+import os
+import os.path as osp
+import sys
+import time
+
+import torch
+
+from torch_geometric import seed_everything
+from torch_geometric.data import download_google_url
+from torch_geometric.datasets import TAGDataset
+from torch_geometric.nn.models import GAT, GCN, GLEM, GraphSAGE
+
+# Add the parent directory to sys.path
+parent_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.append(parent_dir)
+
+
+def get_n_params(model):
+    pp = 0
+    for p in list(model.parameters()):
+        nn = 1
+        for s in list(p.size()):
+            nn = nn * s
+        pp += nn
+    return pp
+
+
+def main(args):
+    gpu = args.gpu
+    dataset_name = args.dataset
+    root = osp.join(parent_dir, 'data', 'ogb')
+    hf_model = args.hf_model
+    pl_ratio = args.pl_ratio
+    gnn_lr = args.gnn_lr
+    lm_lr = args.lm_lr
+    em_order = args.em_order
+    gnn_epochs = args.gnn_epochs
+    lm_epochs = args.lm_epochs
+    patience = args.patience
+    verbose = args.verbose
+    out_dir = args.out_dir
+    lm_batch_size = args.lm_batch_size
+    gnn_batch_size = args.gnn_batch_size
+    lm_use_lora = args.lm_use_lora
+    token_on_disk = args.token_on_disk
+    num_em_iters = args.num_em_iters
+    start_time = time.time()
+    train_without_ext_pred = args.train_without_ext_pred
+    ext_pred = None
+    pretrain_augmented = False
+    ext_pseudo_labels = None
+    device = torch.device(
+        f'cuda:{gpu}' if torch.cuda.is_available() else 'cpu')
+    print(f'Running on: {torch.cuda.get_device_name({gpu})}')
+    torch.cuda.empty_cache()
+
+    if not train_without_ext_pred:
+        ext_pred_path = download_google_url(
+            id='15sO2m7BeW7C1Upmdw3Cx1JS__6nxTAzY',
+            folder='/work/users/junhaos/glem_data/ogbn_products/ext_preds',
+            filename='giant_sagn_scr.pt', log=True)
+        ext_pred = torch.load(ext_pred_path, map_location=device)
+        ext_pseudo_labels = ext_pred.argmax(dim=-1)
+        pretrain_augmented = True
+
+    seed_everything(42)
+    from ogb.nodeproppred import PygNodePropPredDataset
```

Comment:
nit: Let's move the import statement at the start of the file.

Reply:
done, moved to the start of the file